### PR TITLE
ENGINE_load_private_key(): Do foreign key detection without side effects

### LINKED ISSUE #22945: Private key loaded via ENGINE_load_private_key is not usable due to "pkey->pmeth_engine" becomes NULL
Hello.
I have a problem to use private key which is loaded using OpenSSL v3.0.12 engine.
It works well in previous versions including v3.0.11.
Below are error logs.

```
[RSA]
4076AAB67F000000:error:020000B3:rsa routines:rsa_ossl_private_encrypt:missing private key:../openssl-3.0.12/crypto/rsa/rsa_ossl.c:332:
4076AAB67F000000:error:1C880004:Provider routines:rsa_sign:RSA lib:../openssl-3.0.12/providers/implementations/signature/rsa_sig.c:588:
4076AAB67F000000:error:06880006:asn1 encoding routines:ASN1_item_sign_ctx:EVP lib:../openssl-3.0.12/crypto/asn1/a_sign.c:284:
[EC]
40E63E917F000000:error:0800007D:elliptic curve routines:ossl_ecdsa_simple_sign_sig:missing private key:../openssl-3.0.12/crypto/ec/ecdsa_ossl.c:225:
40E63E917F000000:error:06880006:asn1 encoding routines:ASN1_item_sign_ctx:EVP lib:../openssl-3.0.12/crypto/asn1/a_sign.c:284:
```

Follow commit causes the problem.
https://github.com/openssl/openssl/commit/2b74e75331a27fc89cad9c8ea6a26c70019300b5

```e->load_privkey()``` sets engine specific evp methods.
```EVP_PKEY_set1_RSA``` (introduced by above commit) calls ```EVP_PKEY_assign``` and ```EVP_PKEY_set_type```.

``` 
EVP_PKEY *ENGINE_load_private_key(ENGINE *e, const char *key_id,
                                  UI_METHOD *ui_method, void *callback_data)
{
 EVP_PKEY *pkey;
 ...
    // [Reporter] Engine sets its own EVP method which should be called instead of default OpenSSL EVP method
    pkey = e->load_privkey(e, key_id, ui_method, callback_data);
    ...
    /* We enforce check for legacy key */
    switch (EVP_PKEY_get_id(pkey)) {
    case EVP_PKEY_RSA:
        {
        RSA *rsa = EVP_PKEY_get1_RSA(pkey);
        // [Reporter] By calling below, pkey loses its EVP engine information
        EVP_PKEY_set1_RSA(pkey, rsa);		//  [Reporter] Causes problem
        RSA_free(rsa);
        }
        break;
 ...
```
 
Within ```pkey_set_type```, information about EVP methods provided by the engine is invalidated, that is, can't access (call) engine specific EVP methods.
 
```
static int pkey_set_type(EVP_PKEY *pkey, ENGINE *e, int type, const char *str,
                         int len, EVP_KEYMGMT *keymgmt)
{
	...
	// At here, pmeth_engine is reset to NULL
	# ifndef OPENSSL_NO_ENGINE
        /* If we have ENGINEs release them */
        ENGINE_finish(pkey->engine);
        pkey->engine = NULL;
        ENGINE_finish(pkey->pmeth_engine);
        pkey->pmeth_engine = NULL;
	# endif
	...
 
```
 
```EVP_DigestSignInit```
-> ```do_sigver_init```
-> ```EVP_PKEY_CTX_new_from_pkey```
-> ```int_ctx_new```
 
In above example ```pkey->pmeth_engine``` is null, so ```int_ctx_new``` calls OpenSSL default EVP methods.
Because default EVP methods can't handle engine specific pkey, it results in failure.
 
We fully understand that provider is recommend instead of engine, but there still exists various S/W components which use engine.
So kindly address above issue.

### PR DESCRIPTION
**Pull request by t8m**
Doing set1()/get1() has unwanted side effects.
Instead just set the foreign flag appropriately.

Fixes #22945

### CODE REVIEW

**Review by beldmit — DISMISSED** [review]
Hopefully it's less invasive than what I did. LGTM

**Comment by beldmit** [comment]
Approved if fixes #22945

File: crypto/engine/eng_pkey.c [hunk 1]
```diff
@@ -79,48 +80,8 @@ EVP_PKEY *ENGINE_load_private_key(ENGINE *e, const char *key_id,
...
-        break;
-#  ifndef OPENSSL_NO_EC
-    case EVP_PKEY_SM2:
-    case EVP_PKEY_EC:
-        {
-        EC_KEY *ec = EVP_PKEY_get1_EC_KEY(pkey);
-        EVP_PKEY_set1_EC_KEY(pkey, ec);
-        EC_KEY_free(ec);
-        }
-        break;
```
Comment (tmshort) [inline crypto/engine/eng_pkey.c:99]: There is a change in behavior with the new code, `EVP_PKEY_SM2` simply breaks out of the `switch` without doing anything. The original code at least tried to see if it was an EC key.
It also means that `pkey->foreign` is never set for `EVP_PKEY_SM2`.

Comment (beldmit) [reply crypto/engine/eng_pkey.c:99]: Still approved if fixes #22945

**Comment by beldmit** [comment]
Sorry, no. It breaks the problem originally fixed: https://github.com/openssl/openssl/pull/19965#issuecomment-1382889805 

I'm sorry that I approved it before without testing. #22977 seems working.

**Comment by t8m** [comment]
@beldmit I do not think the other approach is right. Instead I would like to know why this patch does not work. Can you please identify why setting the foreign key bit does not work with this patch? IMO we ideally have testcases in the codebase for this and they should have been added with the first fix.

**Comment by beldmit** [comment]
I think I got it. the NSS engine doesn't deal with EVP_PKEY_METHOD, it uses old-style RSA_METHOD/EC_METHOD etc. So the assignment of the low-level object stores the linked RSA_METHOD in the struct.

**Comment by t8m** [comment]
> I think I got it. the NSS engine doesn't deal with EVP_PKEY_METHOD, it uses old-style RSA_METHOD/EC_METHOD etc. So the assignment of the low-level object stores the linked RSA_METHOD in the struct.

Then it needs to actively create the EVP_PKEYs as legacy and not just assume d2i_PUBKEY will produce legacy keys. But I see the current code of e_nss already fixed that.

**Comment by beldmit** [comment]
@t8m how do you run tests? I tried to test your patch with e_nss 4.3 but didn't manage to initialize the database...

**Comment by t8m** [comment]
@beldmit I did not run tests I just looked at the e_nss source in their repo in gitlab.

**Comment by beldmit** [comment]
I managed to run the test, it passes. 
I don't like that we enforce the authors of engines to update their code and lose the attached `*_METHOD`s but as we don't have proper tests, let's live with it.

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Closing in favor of #23063

**Pull request closed without merging**